libnetwork/drivers/bridge/port_mapping_linux.go

```diff
@@ -1,221 +1,318 @@
+// container port, and host port range (their host addresses must differ).
+func bindHostPorts(cfg []*types.PortBinding, proxyPath string) ([]*portBinding, error) {
+	if len(cfg) == 0 {
+		return nil, nil
 	}
 
 	// Try up to maxAllocatePortAttempts times to get a port that's not already allocated.
+	var err error
 	for i := 0; i < maxAllocatePortAttempts; i++ {
-		if host, err = portmapper.MapRange(container, bnd.HostIP, int(bnd.HostPort), int(bnd.HostPortEnd), ulPxyEnabled); err == nil {
-			break
+		var b []*portBinding
+		b, err = attemptBindHostPorts(cfg, cfg[0].Proto.String(), cfg[0].HostPort, cfg[0].HostPortEnd, proxyPath)
+		if err == nil {
+			return b, nil
 		}
 		// There is no point in immediately retrying to map an explicitly chosen port.
-		if bnd.HostPort != 0 && bnd.HostPort == bnd.HostPortEnd {
-			log.G(context.TODO()).Warnf("Failed to allocate and map port %d-%d: %s", bnd.HostPort, bnd.HostPortEnd, err)
+		if cfg[0].HostPort != 0 && cfg[0].HostPort == cfg[0].HostPortEnd {
+			log.G(context.TODO()).Warnf("Failed to allocate and map port: %s", err)
 			break
 		}
 		log.G(context.TODO()).Warnf("Failed to allocate and map port: %s, retry: %d", err, i+1)
 	}
+	return nil, err
+}
+
+// Allow unit tests to supply a dummy StartProxy.
+var startProxy = portmapper.StartProxy
+
+// attemptBindHostPorts allocates host ports for each port mapping that requires
+// one, and reserves those ports by starting docker-proxy.
+//
+// If the allocator doesn't have an available port in the required range, or the
+// docker-proxy process doesn't start (perhaps because another process has
+// already bound the port), all resources are released and an error is returned.
+// When ports are successfully reserved, a portBinding is returned for each
+// mapping.
+func attemptBindHostPorts(
+	cfg []*types.PortBinding,
+	proto string,
+	hostPortStart, hostPortEnd uint16,
+	proxyPath string,
+) (_ []*portBinding, retErr error) {
+	addrs := make([]net.IP, 0, len(cfg))
+	for _, c := range cfg {
+		addrs = append(addrs, c.HostIP)
+	}
+
+	pa := portallocator.Get()
+	port, err := pa.RequestPortsInRange(addrs, proto, int(hostPortStart), int(hostPortEnd))
 	if err != nil {
-		return err
+		return nil, err
 	}
+	defer func() {
+		if retErr != nil {
+			for _, a := range addrs {
+				pa.ReleasePort(a, proto, port)
+			}
+		}
+	}()
 
-	// Save the host port (regardless it was or not specified in the binding)
-	switch netAddr := host.(type) {
-	case *net.TCPAddr:
-		bnd.HostPort = uint16(host.(*net.TCPAddr).Port)
-		return nil
-	case *net.UDPAddr:
-		bnd.HostPort = uint16(host.(*net.UDPAddr).Port)
-		return nil
-	case *sctp.SCTPAddr:
-		bnd.HostPort = uint16(host.(*sctp.SCTPAddr).Port)
-		return nil
-	default:
-		// For completeness
-		return ErrUnsupportedAddressType(fmt.Sprintf("%T", netAddr))
+	res := make([]*portBinding, 0, len(cfg))
+	for _, c := range cfg {
+		pb := &portBinding{PortBinding: c}
```

Comment:
This should copy 'c', so that if startProxy fails the caller will retry with unmodified config.